### PR TITLE
Replace &Option<T> with Option<&T>

### PR DESCRIPTION
**Pull request by gaoxinge**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/4424.

# Rationale for this change

This pr is following: https://github.com/apache/arrow-datafusion/pull/4446, https://github.com/apache/arrow-datafusion/pull/4458 and https://github.com/apache/arrow-datafusion/pull/5102. And I think this is the last pr for closing the issue https://github.com/apache/arrow-datafusion/issues/4424, because it clears all remaining `&Option<T>` type.

# What changes are included in this PR?

Replace `&Option<T>` with `Option<&T>`.

# Are these changes tested?

Pass the original test.

# Are there any user-facing changes?

May change the api.

### CODE REVIEW

File: datafusion/proto/src/logical_plan/from_proto.rs [hunk 1]
```diff
@@ -1389,7 +1415,7 @@ pub fn from_proto_binary_op(op: &str) -> Result<Operator, Error> {
 }
 
 fn parse_optional_expr(
-    p: &Option<Box<protobuf::LogicalExprNode>>,
+    p: Option<Box<protobuf::LogicalExprNode>>,
```
Comment (tustvold) [inline datafusion/proto/src/logical_plan/from_proto.rs:1418]:
```suggestion
    p: Option<&protobuf::LogicalExprNode>,
```

File: datafusion/proto/src/logical_plan/from_proto.rs [hunk 2]
```diff
@@ -876,7 +876,7 @@ pub fn parse_expr(
                 .ok_or_else(|| Error::required("value"))?
                 .try_into()?;
 
-            let expr = parse_required_expr(&field.expr, registry, "expr")?;
+            let expr = parse_required_expr(field.expr.clone(), registry, "expr")?;
```
Comment (tustvold) [inline datafusion/proto/src/logical_plan/from_proto.rs:879]:
```suggestion
            let expr = parse_required_expr(field.expr.as_deref(), registry, "expr")?;
```

File: datafusion/proto/src/logical_plan/from_proto.rs [hunk 3]
```diff
@@ -1399,7 +1425,7 @@ fn parse_optional_expr(
 }
 
 fn parse_required_expr(
-    p: &Option<Box<protobuf::LogicalExprNode>>,
+    p: Option<Box<protobuf::LogicalExprNode>>,
```
Comment (tustvold) [inline datafusion/proto/src/logical_plan/from_proto.rs:1428]:
```suggestion
    p: Option<&protobuf::LogicalExprNode>,
```

**Review by tustvold — COMMENTED** [review]
What do you think about using `as_deref` to avoid needing to clone the expr?

**Review by tustvold — DISMISSED** [review]

**Comment by gaoxinge** [comment]
> What do you think about using `as_deref` to avoid needing to clone the expr?

@tustvold  I change the code from `clone` to `as_deref`.

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
Thank you

**Pull request merged**